### PR TITLE
HDDS-7496. Make default bucket layout FSO in a backwards compatible way.

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

Currently the default value for ozone.default.bucket.layout is LEGACY. This means most users will continue to create legacy buckets. This change makes the default FSO and still allows old clients to create legacy buckets they can read using a pre-process validator to set the bucket type based on the client version.

The current code was setting the bucket layout to legacy for create bucket commands from older clients, but it was not using a pre-process validator as is standard practice. This patch refactors it to do so.

The following Jiras caused test failures related to this change. This patch implements work arounds but we are working on getting them assigned to be properly fixed. Some are actively being fixed already.
- HDDS-5463
    - Recon container -> key mapping does not support FSO
    - PR available
- HDDS-7810
    - Recon DU does not support OBS
- HDDS-7811
    - 500 error with MPU on FSO with bad upload ID.
- HDDS-3402
    - FSO files and directories do not inherit parent bucket acls.

## What is the link to the Apache JIRA

HDDS-7496

## How was this patch tested?

- `TestBucketLayoutWithOlderClient#testCreateBucketWithOlderClient` passes, indicating older clients create legacy buckets regardless of the server default.
- `TestOMBucketCreateRequest#testValidateAndUpdateCacheVerifyBucketLayout` is updated to ensure the default bucket layout is set to FSO.

### CODE REVIEW

**Review by fapifta — COMMENTED** [review]
Hi @errose28 
thank you for taking this up, I think the changes are good, and straightforward. 

The test failures seem to be related to MiniCluster and how it handles the bucket layout for the first sight, so that should be fixed before we can commit though.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/OMBucketCreateRequest.java [hunk 1]
```diff
@@ -153,19 +154,18 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
...
-    // "ozone.default.bucket.layout" configured value for the newer ozone
-    // client and LEGACY for an older ozone client.
+    // bucketInfo.hasBucketLayout() would be true when the request proto
+    // specifies a bucket layout.
+    // When the value is not specified in the proto, OM will use
+    // "ozone.default.bucket.layout".
     if (!bucketInfo.hasBucketLayout()) {
-      BucketLayout defaultBucketLayout =
-          getDefaultBucketLayout(ozoneManager, volumeName, bucketName);
+      BucketLayout defaultBuckLayout = ozoneManager.getOMDefaultBucketLayout();
```
Comment (fapifta) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/OMBucketCreateRequest.java:162]: Is it intentional to change the variable name here or accidental? I am fine with this version also it is pretty much up to you, just wanted to note the change so we are aware.

**Comment by errose28** [comment]
Thanks for checking this out @fapifta. I've updated the PR to properly handle the server side default bucket layout when OM is pre-finalized. The upgrade/downgrade acceptance tests caught this. The test failures are somewhat diverse but most fall in to the following categories:
- Some list and rename key APIs now being used on FSO buckets
- FSO support missing in some areas of Recon
- NPE in the RPC client tests that look to be caused by some sort of test setup issue.

This is going to take some more investigation than I initially thought. I am moving this to draft so I can push test fixes iteratively without using our Apache CI minutes.

**Comment by errose28** [comment]
Another issue is that `ozone.client.fs.default.bucket.layout` defaults to FSO. Even in master right now this means if a bucket is created using ofs while the OM is pre-finalized, it will fail unless this config is set to LEGACY.

If we want to avoid this we could make the fs layout config default to an empty value, which will use the server side default layout if the client side one is unset. This allows the OM to use LEGACY for everything when pre-finalized and switch to using the server default once finalized. This would make the upgrade process smoother, but would mean that if someone set the default server side layout to OBS, they would need to explicitly set this config to FSO otherwise creating a bucket as part of mkdir would fail. IMO this would be better than a surprise during upgrade, and is also similar to what I am currently planning to do on s3 gateway in #3967. However, with 1.3.0 almost out the door and given the scope of test failures in this PR I don't think this will land in the release.

**Review by fapifta — APPROVED** [review]
Thanks for working on this and figuring out various failure scenarios, and a reasonable working model. I think the changes are good to go.

**Comment by errose28** [comment]
I missed the failures in `TestOMRatisSnapshots` which are related to this change. I need to fix them before merging.

**Review by fapifta — APPROVED** [review]

**Comment by errose28** [comment]
Thanks for the review @fapifta

**Pull request merged**